scaffolding-ruby/lib/scaffolding.sh

```diff
@@ -0,0 +1,946 @@
+    exit_with "App type not supported" 2
+    _app_type="rails41"
+    return 0
+  else
+    return 1
+  fi
+}
+
+_detect_rails42_app() {
+  if _has_gem railties && _compare_gem railties \
+      --greater-than-eq 4.2.0 --less-than 5.0.0; then
+    build_line "Detected Rails 4.2 app type"
+    warn "Rails 4.2 app types not yet supported with this Scaffolding"
+    exit_with "App type not supported" 2
+    _app_type="rails42"
+    return 0
+  else
+    return 1
+  fi
+}
+
+_detect_rails5_app() {
+  if _has_gem railties && _compare_gem railties \
+      --greater-than-eq 5.0.0 --less-than 6.0.0; then
+    build_line "Detected Rails 5 app type"
+    _app_type="rails5"
+    return 0
+  else
+    return 1
+  fi
+}
+
+_detect_ruby_app() {
+  build_line "Detected Ruby app type"
+  warn "Ruby app types not yet supported with this Scaffolding"
+  exit_with "App type not supported" 2
+  _app_type="ruby"
+  return 0
+}
+
+_detect_ruby() {
+  local lockfile_version
+
+  if [[ -n "$scaffolding_ruby_pkg" ]]; then
+    _ruby_pkg="$scaffolding_ruby_pkg"
+    build_line "Detected Ruby version in Plan, using '$_ruby_pkg'"
+  else
+    lockfile_version="$($_gemfile_parser ruby-version ./Gemfile.lock || true)"
+    if [[ -n "$lockfile_version" ]]; then
+      # TODO fin: Add more robust Gemfile to Habitat package matching
+      case "$lockfile_version" in
+        *)
+          _ruby_pkg="core/ruby/$(
+            echo "$lockfile_version" | cut -d ' ' -f 2)"
+          ;;
+      esac
+      build_line "Detected Ruby version '$lockfile_version' in Gemfile.lock, using '$_ruby_pkg'"
+    else
+      _ruby_pkg="$_default_ruby_pkg"
+      build_line "No Ruby version detected in Plan or Gemfile.lock, using default '$_ruby_pkg'"
+    fi
+  fi
+  pkg_deps=($_ruby_pkg ${pkg_deps[@]})
+  debug "Updating pkg_deps=(${pkg_deps[*]}) from Scaffolding detection"
+}
+
+_detect_sqlite3() {
+  if _has_gem sqlite3; then
+    build_line "Detected 'sqlite3' gem in Gemfile.lock, adding sqlite packages"
+    pkg_deps=(core/sqlite ${pkg_deps[@]})
+    debug "Updating pkg_deps=(${pkg_deps[*]}) from Scaffolding detection"
+  fi
+}
+
+_detect_webpacker() {
+  if _has_gem webpacker; then
+    build_line "Detected 'webpacker' gem in Gemfile.lock, adding yarn packages"
+    pkg_deps=(core/yarn ${pkg_deps[@]})
+    debug "Updating pkg_deps=(${pkg_deps[*]}) from Scaffolding detection"
+  fi
+}
+
+
+
+
+# **Internal** Invokes the `bundle` program using the chosen version of Ruby.
+# This way, we should avoid most Bundler warnings about mismatched Ruby
+# versions specfied in the Gemfile.
+_bundle() {
+  local bundler_prefix
+  bundler_prefix="$(pkg_path_for bundler)"
+
+  env \
+    -u RUBYOPT \
+    -u GEMRC \
+    GEM_HOME="$bundler_prefix" \
+    GEM_PATH="$bundler_prefix" \
+    "$(pkg_path_for "$_ruby_pkg")/bin/ruby" \
+    "$bundler_prefix/bin/bundle.real" ${*:-}
+}
+
+_rake() {
+  case "$_app_type" in
+    rails*)
+      RACK_ENV=production \
+        RAILS_ENV=production \
```

Comment:
I'd rather not add more environment variables for an app than are strictly necessary if possible. What I've seen in the wild is that this extra context gets copy/pasted around because no one understands why it's there in the first place. I always think of the [ham allegory](http://critical-thinkers.com/2015/01/cutting-off-the-ends-of-the-ham/) in these situations.